File: fix-preview-vscode/src/fixParser.ts
import { SchemaLoader } from './schemaLoader';

export interface FixField {
    tag: string;
    tagName: string;
    value: string;
    enumName: string;
    description: string;
    isRepeatingGroup?: boolean;
    children?: FixField[][];
}

export class FixParser {
    private schemaLoader: SchemaLoader;

    constructor(schemaLoader: SchemaLoader) {
        this.schemaLoader = schemaLoader;
    }

    public parse(raw: string): FixField[] {
        const delimiter = this.detectDelimiter(raw);
        const parts = raw.split(delimiter).filter(p => p.includes('='));
        
        const fields: FixField[] = [];
        const stack: { noTag: string, entries: FixField[][], currentEntry: FixField[] | null }[] = [];

        for (const part of parts) {
            const [tag, value] = part.split('=');
            const metadata = this.schemaLoader.getField(tag);
            const enumMetadata = this.schemaLoader.getEnum(tag, value);

            const field: FixField = {
                tag,
                tagName: metadata?.name || 'Unknown',
                value,
                enumName: enumMetadata?.name || '',
                description: metadata?.description || ''
            };

            // Heuristic for repeating groups: 
            // 1. If tag name starts with "No" and type is NumInGroup (we don't have type here yet, but we can assume from name)
            // 2. If we are in a group and see the first tag of the group again, start new entry.
            // This is complex without full structural schema.
            // Simplified: If tag starts with "No", it's a group header.
            if (field.tagName.startsWith('No') && !isNaN(parseInt(value))) {
                field.isRepeatingGroup = true;
                field.children = [];
                fields.push(field);
                stack.push({ noTag: tag, entries: field.children, currentEntry: null });
                continue;
            }

            if (stack.length > 0) {
                const currentGroup = stack[stack.length - 1];
                // If we see a tag that is already in the current entry, start a new entry
                if (currentGroup.currentEntry && currentGroup.currentEntry.some(f => f.tag === tag)) {
                    currentGroup.currentEntry = [field];
                    currentGroup.entries.push(currentGroup.currentEntry);
                } else if (!currentGroup.currentEntry) {
                    currentGroup.currentEntry = [field];
                    currentGroup.entries.push(currentGroup.currentEntry);
                } else {
                    currentGroup.currentEntry.push(field);
                }
            } else {
                fields.push(field);
            }
        }

        return fields;
    }

    private detectDelimiter(raw: string): string {
        if (raw.includes('\x01')) {
            return '\x01';
        }
        if (raw.includes('|')) {
            return '|';
        }
        return '\x01'; // Default
    }
}

File: fix-preview-intellij/src/main/kotlin/com/antigravity/fixpreview/FixParser.kt
package com.antigravity.fixpreview

data class FixField(
    val tag: String,
    val tagName: String,
    val value: String,
    val enumName: String,
    val description: String,
    val isRepeatingGroup: Boolean = false,
    val section: String = "Body",
    val children: MutableList<MutableList<FixField>>? = null
)

class FixParser(private val schemaLoader: SchemaLoader) {
    private val headerTags = setOf("8", "9", "35", "49", "56", "34", "52", "122", "115", "128", "90", "91", "212", "213", "347", "369", "627", "628", "629", "630")
    private val tailTags = setOf("10", "89", "93")

    fun parse(raw: String): List<FixField> {
        val delimiter = detectDelimiter(raw)
        val parts = raw.split(delimiter).filter { it.contains('=') }

        val fields = mutableListOf<FixField>()
        val stack = mutableListOf<GroupContext>()

        for (part in parts) {
            val keyValue = part.split('=', limit = 2)
            if (keyValue.size < 2) continue
            val tag = keyValue[0]
            val value = keyValue[1]

            val metadata = schemaLoader.getField(tag)
            val enumMetadata = schemaLoader.getEnum(tag, value)

            val isGroupHeader = metadata?.type == "NumInGroup" || (metadata?.name?.startsWith("No") == true && value.toIntOrNull() != null)

            val section = when {
                headerTags.contains(tag) -> "Header"
                tailTags.contains(tag) -> "Tail"
                else -> "Body"
            }

            val field = FixField(
                tag = tag,
                tagName = metadata?.name ?: "Unknown",
                value = value,
                enumName = enumMetadata?.name ?: "",
                description = metadata?.description ?: "",
                isRepeatingGroup = isGroupHeader,
                section = section,
                children = if (isGroupHeader) mutableListOf() else null
            )

            if (isGroupHeader) {
                if (stack.isNotEmpty()) {
                    val currentGroup = stack.last()
                    if (currentGroup.currentEntry == null) {
                        currentGroup.currentEntry = mutableListOf()
                        currentGroup.entries.add(currentGroup.currentEntry!!)
                    }
                    currentGroup.currentEntry!!.add(field)
                } else {
                    fields.add(field)
                }
                stack.add(GroupContext(tag, field.children!!))
                continue
            }

            if (stack.isNotEmpty()) {
                val currentGroup = stack.last()
                // Repeating group logic: if tag already exists in current entry, start new entry
                if (currentGroup.currentEntry != null && currentGroup.currentEntry!!.any { it.tag == tag }) {
                    currentGroup.currentEntry = mutableListOf(field)
                    currentGroup.entries.add(currentGroup.currentEntry!!)
                } else if (currentGroup.currentEntry == null) {
                    currentGroup.currentEntry = mutableListOf(field)
                    currentGroup.entries.add(currentGroup.currentEntry!!)
                } else {
                    currentGroup.currentEntry!!.add(field)
                }
            } else {
                fields.add(field)
            }
        }

        return fields
    }

    private fun detectDelimiter(raw: String): String {
        return if (raw.contains('\u0001')) "\u0001"
        else if (raw.contains('|')) "|"
        else "\u0001"
    }

    private data class GroupContext(
        val noTag: String,
        val entries: MutableList<MutableList<FixField>>,
        var currentEntry: MutableList<FixField>? = null
    )
}
